edi-module/component_writer.bal

```diff
@@ -0,0 +1,137 @@
+            return error Error(string `Mandatory composite field of a segment is not provided. Segment: ${segSchema.tag}, Field: ${fieldSchema.tag}`);
+        } else {
+            return "";
+        }
+    }
+
+    string cd = context.schema.delimiters.component;
+    if componentGroup is map<json> {
+        string[] ckeys = componentGroup.keys();
+        if ckeys.length() < fieldSchema.components.length() && !fieldSchema.truncatable {
+            return error Error(string `Number of components in the composite field does not match with the field schema.
+            Segment ${segSchema.code}, Field ${fieldSchema.tag}, Required components: ${fieldSchema.components.length()}, 
+            Available components: ${ckeys.length()}, Component group value: ${componentGroup.toString()}.`);
+        }
+        int cindex = 0;
+        string cGroupText = "";
+        while cindex < fieldSchema.components.length() {
+            EDIComponentSchema cmap = fieldSchema.components[cindex];
+            if cindex >= ckeys.length() {
+                if cmap.required {
+                    return error Error(string `Mandatory component not found. Segment ${segSchema.code}, Field: ${fieldSchema.tag}, Component: ${cmap.tag}, Component group: ${componentGroup.toString()}`);
+                }
+                cindex += 1;
+                continue;
+            }
+            string ckey = ckeys[cindex];
+            if ckey != cmap.tag {
+                return error Error(string `Component does not match with the schema.
+                Segment: ${segSchema.tag}, Field: ${fieldSchema.tag}, Component schema tag: ${cmap.tag}, Component index: ${cindex}, Component tag: ${ckey}`);
+            }
+            var componentValue = componentGroup.get(ckey);
+            if componentValue is string && componentValue.trim().length() == 0 {
+                if cmap.required {
+                    return error Error(string `Mandatory component not found. Segment: ${segSchema.code}, Field: ${fieldSchema.tag}, Component: ${cmap.tag}, Component group: ${componentGroup.toString()}`);
+                } else {
+                    cGroupText += (cindex == 0 ? "" : cd) + "";
+                    cindex += 1;
+                    continue;
+                }
+            }
+            if cmap.subcomponents.length() == 0 {
+                if componentValue is SimpleType {
+                    cGroupText += (cindex == 0 ? "" : cd) + serializeSimpleType(componentValue, context.schema);
+                    cindex += 1;
+                } else {
+                    return error Error(string `Component must contain a primitive value. Segment: ${segSchema.code}, Field: ${fieldSchema.tag}, Component: ${cmap.tag}, Component value: ${componentValue.toString()}`);
+                }
+            } else if cmap.subcomponents.length() > 0 {
+                string|error scGroupText = writeSubcomponentGroup(componentValue, segSchema, cmap, context);
```

Comment:
for all error checking cases can we check error first and return early to avoid unnecessary else blocks. I think that will simplify the if else ladders.

i.e. 
```ballerina
string|error scGroupText = writeSubcomponentGroup(componentValue, segSchema, cmap, context);
                if scGroupText is error {
                    return error Error(string `Component must contain a composite value. Segment: ${segSchema.code}, Field: ${fieldSchema.tag}, Component: ${cmap.tag}, Component value: ${componentValue.toString()}`);
                } 
                cGroupText += (cindex == 0 ? "" : cd) + scGroupText;
                cindex += 1;
```